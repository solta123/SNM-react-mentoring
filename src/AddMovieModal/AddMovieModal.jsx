import './AddMovieModal.css';
import React from 'react';
import CloseIcon from '@material-ui/icons/Close';
import { TextField } from '@material-ui/core';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import { useFormik } from 'formik';
import { connect } from 'react-redux';
import { genres } from '../common/genres';
import { addMovie, editMovie } from '../store/actionCreator';

const AddMovieModal = React.forwardRef((props, ref) => {
    const formatDate = (date) => {
        return date.getFullYear() + '-' + (date.getMonth() + 1 > 9 ? '' : '0')
            + (date.getMonth() + 1) + '-' + date.getDate()
    }

    const validate = values => {
        const errors = {};

        if (!values.title) {
            errors.title = 'Required title';
        }

        if (!values.release_date) {
            errors.release_date = 'Required release date';
        }

        if (!values.poster_path) {
            errors.poster_path = 'Required to add a link to an image';
        }

        if ((!values.runtime && values.runtime !== 0) || values.runtime < 0) {
            errors.runtime = 'Hmm, this seems a little short...';
        }
        return errors;
    };

    const formik = useFormik({
        initialValues: props.movieDetail?.id ? { ...props.movieDetail } : {
            title: '',
            release_date: formatDate(new Date()),
            genres: [],
            poster_path: '',
            runtime: 0,
            overview: '',
            tagline: '',
            vote_average: 6.0,
            vote_count: 0,
            budget: 0,
            revenue: 0
        },
        validate,
        onSubmit: values => {
            console.log(values)
            props.movieDetail ? props.onEdit(values) : props.onAdd(values)
            props.onCloseModal();
        }
    });

    return <Paper ref={ref} className="ModalContainer">
        <div>
            <CloseIcon onClick={props.onCloseModal} className="closeIcon" />
            <Typography variant="h5">
                {props.movieDetail ? <span>Edit movie</span> : <span>Add movie</span>}
            </Typography>
        </div>

        <form onSubmit={formik.handleSubmit} className="AddMovieForm">
            <div>
                <TextField id="title" name="title" type="text" label="Title"
                    onChange={formik.handleChange} value={formik.values.title} />
            </div>
            {formik.errors.title ? <div className="error">{formik.errors.title}</div> : null}

            <div>
                <TextField label="Release date" InputLabelProps={{ shrink: true }} id="release_date"
                    name="release_date" type="date" onChange={formik.handleChange} value={formik.values.release_date} />
            </div>
            {formik.errors.release_date ? <div className="error">{formik.errors.release_date}</div> : null}

            <div>
                <TextField label="Movie URL" id="poster_path" name="poster_path" type="text"
                    onChange={formik.handleChange} value={formik.values.poster_path}
                />
            </div>
            {formik.errors.poster_path ? <div className="error">{formik.errors.poster_path}</div> : null}

            <div>
                <FormControl>
                    <InputLabel>Genres</InputLabel>
                    <Select id="genres" name="genres" value={formik.values.genres} onChange={formik.handleChange} multiple>
                        {genres.map(genreItem => (
                            <MenuItem key={genreItem} value={genreItem}>{genreItem}</MenuItem>
                        ))}
                    </Select>
                </FormControl>
            </div>

            <div>
                <TextField label="Overview" id="overview" name="overview" type="text"
                    multiline rows={3} rowsMax={10} onChange={formik.handleChange} value={formik.values.overview} />
            </div>

            <div>
                <TextField label="Runtime" id="runtime" name="runtime" type="number"
                    onChange={formik.handleChange} value={formik.values.runtime} />
            </div>
            {formik.errors.runtime ? <div className="error">{formik.errors.runtime}</div> : null}

            <div>
                <Button variant="contained" color="primary" className="AddMovieFormButtons" type="submit">Submit</Button>
                <Button variant="contained" className="AddMovieFormButtons">Reset</Button>
            </div>
        </form>
    </Paper>
});

const mapDispatchToProps = dispatch => {
    return {
        onEdit: (movie) => dispatch(editMovie(movie)),
        onAdd: (movie) => dispatch(addMovie(movie))
    }
}

export default connect(null, mapDispatchToProps)(AddMovieModal);
