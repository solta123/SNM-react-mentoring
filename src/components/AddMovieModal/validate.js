const validate = (values, submitted) => {
    let errors = {};

        if (submitted) {

            if (!values.title) {
                errors.title = 'Required title';
            }

            if (!values.release_date) {
                errors.release_date = 'Required release date';
            }

            if (!values.poster_path) {
                errors.poster_path = 'Required to add a link to an image';
            }

            if (!values.genres.length) {
                errors.genres = 'Please add at least one genre';
            }

            if (!values.overview) {
                errors.overview = 'Required overview';
            }

            if ((!values.runtime && values.runtime !== 0) || values.runtime < 0) {
                errors.runtime = 'Hmm, this seems a little short...';
            }

        } else {
            errors = {
                title: '',
                release_date: '',
                poster_path: '',
                genres: '',
                overview: '',
                runtime: ''
            }
        }

        return errors;
}

export default validate;
