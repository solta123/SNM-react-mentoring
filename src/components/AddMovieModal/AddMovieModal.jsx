import './AddMovieModal.css';
import React, { useState } from 'react';
import CloseIcon from '@material-ui/icons/Close';
import { TextField } from '@material-ui/core';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import { useFormik } from 'formik';
import { connect } from 'react-redux';
import { genres } from '../common/genres';
import { addMovie, editMovie } from '../store/actionCreator';
import { getEmptyMovie, mapMovie } from '../mapper/movieMapper';
import validate from './validate';
import PropTypes from 'prop-types';
import { useTranslation } from "react-i18next";

const AddMovieModal = React.forwardRef((props, ref) => {
    const [submitted, setSubmitted] = useState(false);
    // eslint-disable-next-line no-unused-vars
    const { t, i18n } = useTranslation('common');
    const tSource = 'NETFLIXROULETTE.ADD_MOVIE_MODAL.';

    const formik = useFormik({
        initialValues: {
            ...props.movieDetail ? mapMovie({ ...props.movieDetail }) : getEmptyMovie(),
        },
        validate: values => validate(values, submitted),
        onSubmit: async values => {
            props.movieDetail ? props.onEdit({ ...values, id: props.movieDetail.id }) : props.onAdd(values);
        }
    });

    return <Paper ref={ref} className="ModalContainer">
        <div>
            <CloseIcon onClick={props.onCloseModal} className="closeIcon" />
            <Typography variant="h5">
                {props.movieDetail ? <span>{t(tSource + 'EDIT_MOVIE')}</span> : <span>{t(tSource + 'ADD_MOVIE')}</span>}
            </Typography>
        </div>

        <form onSubmit={formik.handleSubmit} className="AddMovieForm">
            <div>
                <TextField id="title" name="title" type="text" label={t(tSource + 'TITLE')}
                    onChange={formik.handleChange} value={formik.values.title} />
            </div>
            {formik.errors.title ? <div id="title-error" className="error">{t(tSource + formik.errors.title)}</div> : null}

            <div>
                <TextField label={t(tSource + 'RELEASE_DATE')} InputLabelProps={{ shrink: true }} id="release_date"
                    name="release_date" type="date" onChange={formik.handleChange} value={formik.values.release_date} />
            </div>
            {formik.errors.release_date ? <div id="release_date-error" className="error">{t(tSource + formik.errors.release_date)}</div> : null}

            <div>
                <TextField label={t(tSource + 'POSTER_PATH')} id="poster_path" name="poster_path" type="text"
                    onChange={formik.handleChange} value={formik.values.poster_path}
                />
            </div>
            {formik.errors.poster_path ? <div id="poster_path-error" className="error">{t(tSource + formik.errors.poster_path)}</div> : null}

            <div>
                <FormControl>
                    <InputLabel>{t(tSource + 'GENRES')}</InputLabel>
                    <Select id="genres" name="genres" value={formik.values.genres} onChange={formik.handleChange} multiple>
                        {genres.map(genreItem => (
                            <MenuItem id={genreItem} key={genreItem} value={genreItem}>{t('NETFLIXROULETTE.GENRES.' + genreItem.toUpperCase())}</MenuItem>
                        ))}
                    </Select>
                </FormControl>
            </div>
            {formik.errors.genres ? <div id="genres-error" className="error">{t(tSource + formik.errors.genres)}</div> : null}

            <div>
                <TextField label={t(tSource + 'OVERVIEW')} id="overview" name="overview" type="text"
                    multiline rows={3} rowsMax={10} onChange={formik.handleChange} value={formik.values.overview} />
            </div>
            {formik.errors.overview ? <div id="overview-error" className="error">{t(tSource + formik.errors.overview)}</div> : null}

            <div>
                <TextField label={t(tSource + 'RUNTIME')} id="runtime" name="runtime" type="number"
                    onChange={formik.handleChange} value={formik.values.runtime} />
            </div>
            {formik.errors.runtime ? <div id="runtime-error" className="error">{t(tSource + formik.errors.runtime)}</div> : null}

            <div>
                <Button variant="contained" color="primary" className="AddMovieFormButtons" type="submit"
                    id="AddMovieSubmit" onClick={() => setSubmitted(true)}>{t(tSource + 'SUBMIT')}</Button>
                <Button variant="contained" className="AddMovieFormButtons">{t(tSource + 'RESET')}</Button>
            </div>
        </form>
    </Paper>
});

AddMovieModal.propTypes = {
    movieDetail: PropTypes.shape({
        id: PropTypes.number,
        title: PropTypes.string,
        release_date: PropTypes.string,
        poster_path: PropTypes.string,
        genres: PropTypes.array,
        overview: PropTypes.string,
        runtime: PropTypes.string,
        tagline: PropTypes.string,
        budget: PropTypes.number,
        revenue: PropTypes.number,
        vote_average: PropTypes.number,
        vote_count: PropTypes.number
    }),
    onEdit: PropTypes.func,
    onAdd: PropTypes.func,
    onCloseModal: PropTypes.func
}

const mapPropsToState = state => {
    return {
        movieDetail: state.movie.editableMovie
    };
};

const mapDispatchToProps = dispatch => {
    return {
        onEdit: (movie) => dispatch(editMovie(movie)),
        onAdd: (movie) => dispatch(addMovie(movie))
    };
};

export default connect(mapPropsToState, mapDispatchToProps)(AddMovieModal);
