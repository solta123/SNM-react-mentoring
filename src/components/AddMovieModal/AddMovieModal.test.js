import React from "react";
import { mount } from "enzyme";
import AddMovieModal from './AddMovieModal';
import { Provider } from "react-redux";
import { createStore, combineReducers } from 'redux';
import movieReducer from '../store/reducers/movie';
import validate from "./validate";
import renderer from 'react-test-renderer';

const testMovie = {
    title: 'title1',
    poster_path: 'img',
    release_date: '2000-01-01',
    overview: 'asd',
    runtime: 90,
    genres: ['Action'],
    tagline: 'asd'
};

describe("AddMovieModal test", () => {
    let store;
    let ref;
    let wrapper;

    beforeEach(() => {
        store = createStore(
            combineReducers({
                movie: movieReducer
            })
        );
        ref = React.createRef();
        wrapper = mount(
            <Provider store={store} >
                <AddMovieModal ref={ref} />
            </Provider>
        );
    });

    afterEach(() => {
        wrapper.unmount();
    })

    it('renders correctly', () => {
        expect(wrapper).toBeTruthy();
    });

    it('error messages appear', () => {
        wrapper.find('button#AddMovieSubmit').simulate('click');
        const errorMessages = wrapper.find('div.error');
        expect(errorMessages).toBeTruthy();
    });

    it('should runtime message appear if empty', () => {
        wrapper.find('input#title').simulate('keydown', { key: 'a' });
        wrapper.find('input#poster_path').simulate('keydown', { key: 'a' });
        wrapper.find('textarea#overview').simulate('keydown', { key: 'a' });
        wrapper.find('input#release_date')
            .simulate('keydown', { key: '2' })
            .simulate('keydown', { key: '0' })
            .simulate('keydown', { key: '0' })
            .simulate('keydown', { key: '0' })
            .simulate('keydown', { key: '-' })
            .simulate('keydown', { key: '0' })
            .simulate('keydown', { key: '1' })
            .simulate('keydown', { key: '-' })
            .simulate('keydown', { key: '0' })
            .simulate('keydown', { key: '1' });
        wrapper.find('input#runtime').simulate('keydown', { key: '9' });
        wrapper.find('div#genres').simulate('change', { target: { value: ['Action'] } });

        wrapper.find('button#AddMovieSubmit').simulate('click');
        expect(wrapper.find('div.error').exists()).toBeFalsy();
        expect(renderer.create(wrapper.text()).toJSON()).toMatchSnapshot();
    });

    it('should fill inputs with data when editing', () => {
        store.getState().movie.editableMovie = { ...testMovie };
        const component = mount(
            <Provider store={store}>
                <AddMovieModal ref={ref} movieDetail={testMovie} />
            </Provider>
        );

        expect(component.find('input#title').html()).toContain('value="title1"');
        expect(component.find('input#release_date').html()).toContain('value="2000-01-01"');
        expect(component.find('textarea#overview').html()).toContain('asd');
        expect(component.find('input#runtime').html()).toContain('value="90"');
        expect(component.find('div#genres').html()).toContain('Action');
        component.unmount();
    });

    it('should not display error messagess if not submitted', () => {
        wrapper.find('input#title').simulate('keydown', { key: 'a' });
        expect(wrapper.find('div.error').exists()).toBe(false);
    });

    it('should validate the form and set errors', () => {
        const errors = validate({ genres: [] }, true);
        expect(errors).toEqual({
            title: 'Required title',
            release_date: 'Required release date',
            poster_path: 'Required to add a link to an image',
            genres: 'Please add at least one genre',
            overview: 'Required overview',
            runtime: 'Hmm, this seems a little short...'
        });
        const errors2 = validate({ genres: [], runtime: -2 }, true);
        expect(errors2).toEqual({
            title: 'Required title',
            release_date: 'Required release date',
            poster_path: 'Required to add a link to an image',
            genres: 'Please add at least one genre',
            overview: 'Required overview',
            runtime: 'Hmm, this seems a little short...'
        });
    });

    it('should not validate when it is not submitted', () => {
        const errors = validate({}, false);
        expect(errors).toEqual({
            title: '',
            release_date: '',
            poster_path: '',
            genres: '',
            overview: '',
            runtime: ''
        });
    });
});
