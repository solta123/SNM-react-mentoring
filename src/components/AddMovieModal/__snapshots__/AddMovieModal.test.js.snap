// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMovieModal test should runtime message appear if empty 1`] = `"Add movieTitleRelease dateMovie URLGenresâ€‹OverviewRuntimeSubmitReset"`;
